pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,905 @@
+	if del, deleted := obj.(cache.DeletedFinalStateUnknown); deleted {
+		controller.volumeIndex.Delete(del.Obj)
+		return
+	}
+	if _, ok := obj.(*api.PersistentVolume); ok {
+		controller.volumeIndex.Delete(obj)
+	}
+}
+
+func (controller *PersistentVolumeController) handleAddClaim(obj interface{}) {
+	controller.locks["_main"].Lock()
+	defer controller.locks["_main"].Unlock()
+	cachedPvc, exists, _ := controller.claimStore.Get(obj)
+	if !exists {
+		glog.Errorf("PersistentVolumeCache does not exist in the local cache: %#v", obj)
+		return
+	}
+	if pvc, ok := cachedPvc.(*api.PersistentVolumeClaim); ok {
+		_, reconciledStatus, err := controller.reconcileClaim(pvc)
+		if err != nil {
+			glog.Errorf("Error encoutered reconciling claim %s: %#v", pvc.Name, err)
+		}
+		if pvc.Status.Phase != reconciledStatus.Phase {
+			pvc.Status = reconciledStatus
+			controller.client.UpdatePersistentVolumeClaimStatus(pvc)
+		}
+		glog.V(5).Infof("PersistentVolumeClaim[%s] is reconciled", pvc.Name)
+	}
+}
+
+func (controller *PersistentVolumeController) handleUpdateClaim(oldObj, newObj interface{}) {
+	controller.handleAddClaim(newObj)
+}
+
+func (controller *PersistentVolumeController) handleDeleteClaim(obj interface{}) {
+	controller.locks["_main"].Lock()
+	defer controller.locks["_main"].Unlock()
+	if del, deleted := obj.(cache.DeletedFinalStateUnknown); deleted {
+		controller.claimStore.Delete(del.Obj)
+		return
+	}
+	if _, ok := obj.(*api.PersistentVolumeClaim); ok {
+		controller.claimStore.Delete(obj)
+	}
+}
+
+func (controller *PersistentVolumeController) reconcileClaim(claim *api.PersistentVolumeClaim) (*api.PersistentVolumeClaim, api.PersistentVolumeClaimStatus, error) {
+	glog.V(5).Infof("Reconciling PersistentVolumeClaim[%s]", claim.Name)
+	// claims can be in one of the following phases
+	//
+	//    ClaimPending -- default value -- not bound to a volume. A volume matching the claim may not exist.
+	//    ClaimBound -- claim is bound to a volume and the volume exists.
+	//    TODO: needs ClaimError phase/message like pv.status
+
+	// Bound
+	if claim.Spec.VolumeName != "" {
+		glog.V(5).Infof("PVC[%s] is bound to %s", claim.Name, claim.Spec.VolumeName)
+		obj, exists, _ := controller.volumeStore.GetByKey(claim.Spec.VolumeName)
+		if !exists {
+			glog.V(5).Infof("PersistentVolumeClaim[%s] could not find bound volume", claim.Name)
+			return claim, api.PersistentVolumeClaimStatus{Phase: api.ClaimPending}, fmt.Errorf("PersistentVolumeClaim[%s] could not find bound volume: %s", claim.Spec.VolumeName)
+		}
+		pv, _ := obj.(*api.PersistentVolume)
+		if pv.Spec.ClaimRef == nil {
+			glog.V(5).Infof("PVC[%s] - found volume to match claim's bind - %s", claim.Name, pv.Name)
+			claimRef, err := api.GetReference(claim)
+			if err != nil {
+				return claim, api.PersistentVolumeClaimStatus{
+					Phase:       api.ClaimPending,
+					Capacity:    pv.Spec.Capacity,
+					AccessModes: pv.Spec.AccessModes,
+				}, fmt.Errorf("Error getting claim reference for bind")
```

Comment:
Looking at the rest of the code, some errors returned from reconcile* methods have claim.Name and some not. Without judging which one is better, one approach should be used. Does the caller add pv/pvc.name to any log messages?